Add support for prefix configuration in the CLI config class.

danielpalstra opened the pull request:
This change adds a prefix configration to the jenkins::cli::config class. The prefix is then used in the CLI command to connect to a Jenkins servers which has a prefix configured (in the config_hash).

jhoblitt commented:
@danielpalstra Instead of interpolating fragments of the jenkins URL, I think it would be more flexible to be able to specify the entire URL.  What do you think?

danielpalstra commented:
@jhoblitt personally I like the setup that the "old" CLI uses. For example  https://github.com/jenkinsci/puppet-jenkins/blob/master/lib/puppet/parser/functions/jenkins_port.rb#L16 takes the port from the config_hash which I think is very clean since you only have to define the port once. Same applies for the prefix. I would consider this a long term solution for the cli config as a whole. 

Regarding interpolating the Jenkins URL. Can you think of any use cases where a different host is used than localhost? I also wouldn't expect that people connect to the CLI by using a https url. So interpolating the url feels fine by me. Can you share your thoughts on this one as well?

danielpalstra commented:
@jhoblitt what do you think of my proposal?

bjwschaap commented:
+1 
We need this as well. I like this particular solution since it is backwards-compatible and has very little impact. Interpolating the URL could be done in a separate enhancement request..

please merge

superseb commented:
+1

jhoblitt commented:
I still feel that removal of the `port` parameter completely and replacing it with a canonical URL would be a better solution.  Backwards compatibility is not a strong concern with this code, the API is documented as being unstable.

danielpalstra commented:
I assume you mean the `prefix` and `port` parameter. I will remove those and create a parameter that contains the complete url of the Jenkins node that the cli needs to connect to.

danielpalstra commented:
@jhoblitt as discussed I've removed the `port` param and replaced it with a canonical `url` param. This had impact on the docs and spec tests which I've updated as well.

jhoblitt commented on NATIVE_TYPES_AND_PROVIDERS.md:
````diff
@@ -48,11 +48,19 @@ class { 'jenkins::cli::config':
+
+```
+class { 'jenkins::cli::config':
+  ssh_private_key => 'http://localhost:9999/awesome-jenkins',
````
That is unlikely to work...

jhoblitt commented:
@danielpalstra This looks good (other then the README typo)!  I'm going to hold on merging this until I have a change to try it but it should go in soon.

danielpalstra commented:
@jhoblitt sounds good. If you like I can send in a gist containing some sample code to test.

jhoblitt commented:
I've restarted travis.  It is starting to look like puppet 4.3 / ruby 1.9.3 may be a broken combo.

danielpalstra commented:
Jup, keeps on failing :)

jhoblitt commented:
A few of the other ruby 1.9.3 combos are also running for over 10 minutes.  Travis will kill a job if there is no output for 10 minutes.  I'm not sure why the ruby 1.9.3 runtimes are so much longer.

jhoblitt commented:
@danielpalstra The ruby 1.9.3/puppet 4.3 combo failed again and we are seeing this with other PRs as well.  I'm going to merge this.  Thank you for your patience!
